### PR TITLE
Extend configuration to set nonce mechanism

### PR DESCRIPTION
**Pull request by weilbith**
Closes: #362

I'm not too happy how the tests look like. Open for suggestions.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +25,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Add function `Payment.confirmPayment` to confirm any `PaymentTxObject` returned by `prepare`
   and potentially send a payment message along with the payment
 - Add function `Messaging.paymentMessage` to send a payment message for a messageId to a counterparty address
+- Add optional field to the `TLNetworkConfig` for choosing the to use nonce mechanism
```
Comment (cducrest) [inline CHANGELOG.md:28]: `the nonce mechanism to use`

File: src/wallets/IdentityWallet.ts [hunk 2]
```diff
@@ -43,16 +44,18 @@ export class IdentityWallet implements TLWallet {
...
   private identityImplementationAddress: string
+  private nonceMechanism: NonceMechanism
   private chainId: number
   private identityVersion = 1
 
   constructor(
     provider: TLProvider,
     chainId: number,
-    { identityFactoryAddress, identityImplementationAddress }
+    { identityFactoryAddress, identityImplementationAddress, nonceMechanism }
```
Comment (cducrest) [inline src/wallets/IdentityWallet.ts:54]: You're right, I am not sure this way of writing inputs makes sense. I would rather have them not packed with proper typing

File: tests/unit/IdentityWallet.test.ts [hunk 3]
```diff
@@ -230,6 +231,60 @@ describe('unit', () => {
...
+
+      beforeEach(() => init())
+
+      it('should generate nonce with random mechanism', async () => {
+        const randomIdentityWallet = new IdentityWallet(
+          fakeTLProvider,
+          FAKE_CHAIN_ID,
+          randomNonceConfig
+        )
+        console.log(await randomIdentityWallet)
```
Comment (cducrest) [inline tests/unit/IdentityWallet.test.ts:261]: remove log

File: CHANGELOG.md (hunk 1 above)
Comment (weilbith) [reply CHANGELOG.md]: :heavy_check_mark:

File: src/wallets/IdentityWallet.ts (hunk 2 above)
Comment (weilbith) [reply src/wallets/IdentityWallet.ts]: :heavy_check_mark:

File: tests/unit/IdentityWallet.test.ts (hunk 3 above)
Comment (weilbith) [reply tests/unit/IdentityWallet.test.ts]: :heavy_check_mark:

**Comment by cducrest** [comment]
I don't get how this test is passing: https://github.com/trustlines-protocol/clientlib/blob/eb27d2f9333f6ec3a3fc8c330ca623b994626a6f/tests/e2e/Identity.test.ts#L92

**Comment by weilbith** [comment]
Hmm. Why? We use the `firstNone`. So it should actually increase shouldn't it?

File: src/TLNetwork.ts [hunk 4]
```diff
@@ -264,15 +266,18 @@ export class TLNetwork {
     walletType: string,
     provider: TLProvider,
     chainId: number,
-    { identityFactoryAddress, identityImplementationAddress }
+    { identityFactoryAddress, identityImplementationAddress, nonceMechanism }
```
Comment (cducrest) [inline src/TLNetwork.ts:269]: You could unpack it here too

Comment (weilbith) [reply src/TLNetwork.ts]: :heavy_check_mark:

**Comment by weilbith** [comment]
The PR is done. I'll rebase it to include the fixup commit before merging it.

**Pull request merged**